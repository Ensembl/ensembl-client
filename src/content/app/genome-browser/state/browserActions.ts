/**
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { createAction } from 'typesafe-actions';
import { Dispatch, ActionCreator, Action } from 'redux';
import { replace } from 'connected-react-router';
import { ThunkAction } from 'redux-thunk';
import isEqual from 'lodash/isEqual';
import pickBy from 'lodash/pickBy';

import * as urlFor from 'src/shared/helpers/urlHelper';
import { getChrLocationStr } from '../helpers/browserHelper';
import { buildFocusIdForUrl } from 'src/shared/helpers/focusObjectHelpers';

import browserStorageService from '../services/browser-storage-service';
import trackPanelStorageService from 'src/content/app/genome-browser/components/track-panel/services/track-panel-storage-service';

import { fetchEnsObject } from 'src/shared/state/focus-object/focusObjectSlice';
import { ensureSpeciesIsEnabled } from 'src/content/app/species-selector/state/speciesSelectorSlice';
import {
  getBrowserActiveGenomeId,
  getBrowserActiveEnsObjectId,
  getBrowserTrackStates,
  getChrLocation,
  getBrowserActiveEnsObjectIds,
  getBrowserNavOpenState
} from './browserSelectors';

import { RootState } from 'src/store';
import { BrowserTrackStates } from '../components/track-panel/trackPanelConfig';
import {
  BrowserNavIconStates,
  ChrLocation,
  CogList,
  ChrLocations
} from './browserState';
import { TrackActivityStatus } from 'src/content/app/genome-browser/components/track-panel/trackPanelConfig';

export type UpdateTrackStatesPayload = {
  genomeId: string;
  categoryName: string;
  trackId: string;
  status: TrackActivityStatus;
};

export type ParsedUrlPayload = {
  activeGenomeId: string;
  activeEnsObjectId: string | null;
  chrLocation: ChrLocation | null;
};

export const setDataFromUrl = createAction(
  'browser/set-data-from-url'
)<ParsedUrlPayload>();

export const setDataFromUrlAndSave: ActionCreator<
  ThunkAction<void, any, null, Action<string>>
> = (payload: ParsedUrlPayload) => (dispatch) => {
  dispatch(setDataFromUrl(payload));

  const { activeGenomeId, activeEnsObjectId, chrLocation } = payload;

  dispatch(ensureSpeciesIsEnabled(payload.activeGenomeId));
  browserStorageService.saveActiveGenomeId(payload.activeGenomeId);
  chrLocation &&
    browserStorageService.updateChrLocation({ [activeGenomeId]: chrLocation });

  if (activeEnsObjectId) {
    browserStorageService.updateActiveEnsObjectIds({
      [activeGenomeId]: activeEnsObjectId
    });
  }
};

export const setActiveGenomeId = createAction(
  'browser/set-active-genome-id'
)<string>();

export const fetchDataForLastVisitedObjects: ActionCreator<
  ThunkAction<void, any, null, Action<string>>
> = () => async (dispatch, getState: () => RootState) => {
  const state = getState();
  const activeEnsObjectIdsMap = getBrowserActiveEnsObjectIds(state);
  Object.values(activeEnsObjectIdsMap).forEach((objectId) =>
    dispatch(fetchEnsObject(objectId))
  );
};

export const updateBrowserActiveEnsObjectIds = createAction(
  'browser/update-active-ens-object-ids'
)<{ [objectId: string]: string }>();

export const updateBrowserActiveEnsObjectIdsAndSave = (
  activeEnsObjectId: string
): ThunkAction<void, any, null, Action<string>> => {
  return (dispatch, getState: () => RootState) => {
    const state = getState();
    const activeGenomeId = getBrowserActiveGenomeId(state);
    if (!activeGenomeId) {
      return;
    }
    const currentActiveEnsObjectIds = getBrowserActiveEnsObjectIds(state);
    const updatedActiveEnsObjectIds = {
      ...currentActiveEnsObjectIds,
      [activeGenomeId]: activeEnsObjectId
    };

    dispatch(updateBrowserActiveEnsObjectIds(updatedActiveEnsObjectIds));
    dispatch(fetchEnsObject(activeEnsObjectId));

    browserStorageService.updateActiveEnsObjectIds(updatedActiveEnsObjectIds);
  };
};

export const updateDefaultPositionFlag = createAction(
  'browser/update-default-position-flag'
)<boolean>();

export const updateTrackStates = createAction(
  'browser/update-tracks-state'
)<BrowserTrackStates>();

export const updateTrackStatesAndSave: ActionCreator<
  ThunkAction<void, any, null, Action<string>>
> = (payload: BrowserTrackStates) => (dispatch, getState: () => RootState) => {
  dispatch(updateTrackStates(payload));
  const trackStates = getBrowserTrackStates(getState());
  browserStorageService.saveTrackStates(trackStates);
};

export const openBrowserNav = createAction('browser/open-browser-navigation')<{
  activeGenomeId: string;
}>();

export const closeBrowserNav = createAction(
  'browser/close-browser-navigation'
)<{ activeGenomeId: string }>();

export const toggleBrowserNav: ActionCreator<
  ThunkAction<any, any, null, Action<string>>
> = () => {
  return (dispatch: Dispatch, getState: () => RootState) => {
    const state = getState();
    const isBrowserNavOpenState = getBrowserNavOpenState(state);
    const activeGenomeId = getBrowserActiveGenomeId(state);

    if (!activeGenomeId) {
      return;
    }
    if (isBrowserNavOpenState) {
      dispatch(closeBrowserNav({ activeGenomeId }));
    } else {
      dispatch(openBrowserNav({ activeGenomeId }));
    }
  };
};

export const updateBrowserNavIconStates = createAction(
  'browser/update-browser-nav-states'
)<{ activeGenomeId: string; navStates: BrowserNavIconStates }>();

export const updateChrLocation = createAction(
  'browser/update-chromosome-location'
)<ChrLocations>();

export const updateActualChrLocation = createAction(
  'browser/update-actual-chromosome-location'
)<ChrLocations>();

export const setActualChrLocation: ActionCreator<
  ThunkAction<any, any, null, Action<string>>
> = (chrLocation: ChrLocation) => {
  return (dispatch: Dispatch, getState: () => RootState) => {
    const state = getState();
    const activeGenomeId = getBrowserActiveGenomeId(state);
    if (!activeGenomeId) {
      return;
    }
    const payload = {
      [activeGenomeId]: chrLocation
    };

    dispatch(updateActualChrLocation(payload));
  };
};

export const setChrLocation: ActionCreator<
  ThunkAction<any, any, null, Action<string>>
> = (chrLocation: ChrLocation) => {
  return (dispatch: Dispatch, getState: () => RootState) => {
    const state = getState();
    const activeGenomeId = getBrowserActiveGenomeId(state);
    const activeEnsObjectId = getBrowserActiveEnsObjectId(state);
    const savedChrLocation = getChrLocation(state);
    if (!activeGenomeId || !activeEnsObjectId) {
      return;
    }
    const payload = {
      [activeGenomeId]: chrLocation
    };

    dispatch(updateChrLocation(payload));
    browserStorageService.updateChrLocation(payload);

    if (!isEqual(chrLocation, savedChrLocation)) {
      const newUrl = urlFor.browser({
        genomeId: activeGenomeId,
        focus: buildFocusIdForUrl(activeEnsObjectId),
        location: getChrLocationStr(chrLocation)
      });
      dispatch(replace(newUrl));
    }
  };
};

export const updateCogList = createAction('browser/update-cog-list')<number>();

export const updateCogTrackList = createAction(
  'browser/update-cog-track-list'
)<CogList>();

export const updateSelectedCog = createAction('browser/update-selected-cog')<
  string | null
>();

export const updateTrackConfigNames = createAction(
  'browser/update-track-config-names',
  (selectedCog: string, sense: boolean) =>
    [selectedCog, sense] as [string, boolean]
)();

export const updateTrackConfigLabel = createAction(
  'browser/update-track-config-label',
  (selectedCog: string, sense: boolean) =>
    [selectedCog, sense] as [string, boolean]
)();

export const updateApplyToAll = createAction(
  'browser/update-apply-to-all'
)<boolean>();

export const updateApplyToAllTrackNames = createAction(
  'browser/update-apply-to-all-track-names'
)<boolean>();

export const updateApplyToAllTrackLabels = createAction(
  'browser/update-apply-to-all-track-labels'
)<boolean>();

export const toggleRegionEditorActive = createAction(
  'browser/toggle-region-editor-active'
)<boolean>();

export const toggleRegionFieldActive = createAction(
  'browser/toggle-region-field-active'
)<boolean>();

export const deleteGenome = createAction('browser/delete-genome')<string>();

export const deleteSpeciesInGenomeBrowser = (
  genomeIdToRemove: string
): ThunkAction<void, any, null, Action<string>> => {
  return (dispatch, getState: () => RootState) => {
    const state = getState();

    dispatch(deleteGenome(genomeIdToRemove));

    const updatedActiveEnsObjectIds = pickBy(
      getBrowserActiveEnsObjectIds(state),
      (value, key) => key !== genomeIdToRemove
    );

    dispatch(updateBrowserActiveEnsObjectIds(updatedActiveEnsObjectIds));

    browserStorageService.deleteGenome(genomeIdToRemove);
    trackPanelStorageService.deleteGenome(genomeIdToRemove);
  };
};
