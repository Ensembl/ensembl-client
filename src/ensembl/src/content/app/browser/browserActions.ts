import { createAction } from 'typesafe-actions';
import { Dispatch, ActionCreator, Action } from 'redux';
import { ThunkAction } from 'redux-thunk';

import config from 'config';

import { BrowserNavStates, ChrLocation, CogList } from './browserState';
import {
  getBrowserActiveGenomeId,
  getBrowserActiveEnsObjectId,
  getDefaultChrLocation,
  getChrLocation
} from './browserSelectors';
import { getBrowserAnalyticsObject } from 'src/analyticsHelper';
import browserStorageService from './browser-storage-service';
import { RootState } from 'src/store';

export const updateBrowserActivated = createAction(
  'browser/update-browser-activated',
  (resolve) => {
    return (browserActivated: boolean) =>
      resolve(browserActivated, getBrowserAnalyticsObject('Default Action'));
  }
);

export const activateBrowser = (browserEl: HTMLDivElement) => {
  return (dispatch: Dispatch) => {
    const { protocol, host: currentHost } = location;
    const host = config.apiHost || currentHost;
    const activateEvent = new CustomEvent('bpane-activate', {
      bubbles: true,
      detail: {
        'config-url': `${protocol}${host}/browser/config`,
        key: 'main'
      }
    });

    browserEl.dispatchEvent(activateEvent);

    dispatch(updateBrowserActivated(true));
  };
};

export const updateBrowserActiveGenomeId = createAction(
  'browser/update-active-genome-id',
  (resolve) => {
    return (activeGenomeId: string) =>
      resolve(activeGenomeId, getBrowserAnalyticsObject('Navigation'));
  }
);

export const updateBrowserActiveGenomeIdAndSave: ActionCreator<
  ThunkAction<void, any, null, Action<string>>
> = (activeGenomeId: string) => (dispatch: Dispatch) => {
  dispatch(updateBrowserActiveGenomeId(activeGenomeId));
  browserStorageService.saveActiveGenomeId(activeGenomeId);
};

export const updateBrowserActiveEnsObjectId = createAction(
  'browser/update-active-ens-object-id',
  (resolve) => {
    return (activeEnsObjectId: { [objectId: string]: string }) =>
      resolve(activeEnsObjectId, getBrowserAnalyticsObject('Navigation'));
  }
);

export const updateBrowserActiveEnsObjectIdAndSave: ActionCreator<
  ThunkAction<void, any, null, Action<string>>
> = (activeEnsObjectId: string) => {
  return (dispatch: Dispatch, getState: () => RootState) => {
    const currentActiveEnsObjectId = getBrowserActiveEnsObjectId(getState());
    const updatedActiveEnsObjectId = { ...currentActiveEnsObjectId };
    const genomeId = activeEnsObjectId.split(':')[0];
    updatedActiveEnsObjectId[genomeId] = activeEnsObjectId;

    dispatch(updateBrowserActiveEnsObjectId(updatedActiveEnsObjectId));

    browserStorageService.updateActiveEnsObjectId(updatedActiveEnsObjectId);
  };
};

export const toggleBrowserNav = createAction(
  'browser/toggle-browser-navigation',
  (resolve) => {
    return () => resolve(undefined, getBrowserAnalyticsObject('Navigation'));
  }
);

export const updateBrowserNavStates = createAction(
  'browser/update-browser-nav-states',
  (resolve) => {
    return (browserNavStates: BrowserNavStates) =>
      resolve(browserNavStates, getBrowserAnalyticsObject('Navigation'));
  }
);

export const updateChrLocation = createAction(
  'browser/update-chromosome-location',
  (resolve) => {
    return (chrLocationData: { [genomeId: string]: ChrLocation }) =>
      resolve(chrLocationData);
  }
);

export const setChrLocation: ActionCreator<
  ThunkAction<any, any, null, Action<string>>
> = (chrLocation: ChrLocation) => {
  return (dispatch: Dispatch, getState: () => RootState) => {
    const genomeId = getBrowserActiveGenomeId(getState());
    const currentChrLocation = getChrLocation(getState());
    const updatedChrLocation = { ...currentChrLocation };
    updatedChrLocation[genomeId] = chrLocation;

    dispatch(updateChrLocation(updatedChrLocation));
    browserStorageService.updateChrLocation(updatedChrLocation);
  };
};

export const updateDefaultChrLocation = createAction(
  'browser/update-default-chromosome-location',
  (resolve) => {
    return (chrLocationData: { [genomeId: string]: ChrLocation }) =>
      resolve(chrLocationData, getBrowserAnalyticsObject('User Interaction'));
  }
);

export const changeBrowserLocation: ActionCreator<
  ThunkAction<any, any, null, Action<string>>
> = (chrLocation: ChrLocation, browserEl: HTMLDivElement) => {
  return (dispatch: Dispatch, getState: () => RootState) => {
    const [chrCode, startBp, endBp] = chrLocation;
    const genomeId = getBrowserActiveGenomeId(getState());

    const stickEvent = new CustomEvent('bpane', {
      bubbles: true,
      detail: {
        stick: `${genomeId}:${chrCode}`
      }
    });

    browserEl.dispatchEvent(stickEvent);

    if (startBp > 0 && endBp > 0) {
      const gotoEvent = new CustomEvent('bpane', {
        bubbles: true,
        detail: {
          goto: `${startBp}-${endBp}`
        }
      });

      browserEl.dispatchEvent(gotoEvent);
    }

    const currentChrLocation = getChrLocation(getState());
    const updatedChrLocation = { ...currentChrLocation };
    updatedChrLocation[genomeId] = chrLocation;

    dispatch(updateChrLocation(updatedChrLocation));
    browserStorageService.updateChrLocation(updatedChrLocation);

    const currentDefaultChrLocation = getDefaultChrLocation(getState());

    if (!currentDefaultChrLocation[genomeId]) {
      const updatedDefaultChrLocation = { ...currentDefaultChrLocation };
      updatedDefaultChrLocation[genomeId] = chrLocation;
      dispatch(updateDefaultChrLocation(updatedDefaultChrLocation));
      browserStorageService.updateDefaultChrLocation(updatedDefaultChrLocation);
    }
  };
};

export const updateCogList = createAction(
  'browser/update-cog-list',
  (resolve) => {
    return (cogList: number) => {
      return resolve(cogList, getBrowserAnalyticsObject('User Interaction'));
    };
  }
);

export const updateCogTrackList = createAction(
  'browser/update-cog-track-list',
  (resolve) => {
    return (trackY: CogList) => {
      return resolve(trackY, getBrowserAnalyticsObject('User Interaction'));
    };
  }
);

export const updateSelectedCog = createAction(
  'browser/update-selected-cog',
  (resolve) => {
    return (index: string) => {
      return resolve(index, getBrowserAnalyticsObject('User Interaction'));
    };
  }
);

export const updateTrackConfigNames = createAction(
  'browser/update-track-config-names',
  (resolve) => {
    return (selectedCog: any, sense: boolean) => {
      return resolve(
        [selectedCog, sense],
        getBrowserAnalyticsObject('User Interaction')
      );
    };
  }
);

export const updateTrackConfigLabel = createAction(
  'browser/update-track-config-label',
  (resolve) => {
    return (selectedCog: any, sense: boolean) => {
      return resolve(
        [selectedCog, sense],
        getBrowserAnalyticsObject('User Interaction')
      );
    };
  }
);

export const updateApplyToAll = createAction(
  'browser/update-apply-to-all',
  (resolve) => {
    return (yn: boolean) => {
      return resolve(yn, getBrowserAnalyticsObject('User Interaction'));
    };
  }
);

export const toggleGenomeSelector = createAction(
  'toggle-genome-selector',
  (resolve) => {
    return (genomeSelectorActive: boolean) =>
      resolve(
        genomeSelectorActive,
        getBrowserAnalyticsObject('User Interaction')
      );
  }
);
