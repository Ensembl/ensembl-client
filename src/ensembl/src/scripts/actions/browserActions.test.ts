import configureStore from 'redux-mock-store';

import * as browser from './browserActions';

const mockStore = configureStore();
const store = mockStore();

describe('browserActions', () => {
  beforeEach(() => {
    store.clearActions();
  });

  test('dispatches toggle actions correctly', () => {
    const expectedActions = [
      {
        type: 'browser/toggle-track-panel'
      },
      {
        type: 'browser/toggle-drawer'
      },
      {
        type: 'browser/open-drawer'
      },
      {
        type: 'browser/close-drawer'
      }
    ];

    store.dispatch(browser.toggleTrackPanel());
    store.dispatch(browser.toggleDrawer());
    store.dispatch(browser.openDrawer());
    store.dispatch(browser.closeDrawer());

    expect(store.getActions()).toEqual(expectedActions);
  });

  test('dispatches changeCurrentTrack correctly', () => {
    const expectedActions = [
      {
        payload: 'track-1',
        type: 'browser/change-current-track'
      }
    ];

    store.dispatch(browser.changeCurrentTrack('track-1'));
    expect(store.getActions()).toEqual(expectedActions);
  });

  test('dispatches changeCurrentDrawerSection correctly', () => {
    const expectedActions = [
      {
        payload: 'summary',
        type: 'browser/change-current-drawer-section'
      }
    ];

    store.dispatch(browser.changeCurrentDrawerSection('summary'));
    expect(store.getActions()).toEqual(expectedActions);
  });

  test('dispaches the correct actions and payloads', () => {
    store.dispatch(browser.toggleTrackPanel());
    store.dispatch(browser.toggleDrawer());
    store.dispatch(browser.openDrawer());
    store.dispatch(browser.closeDrawer());
    store.dispatch(browser.changeCurrentTrack('track-1'));
    store.dispatch(browser.changeCurrentDrawerSection('main'));

    expect(store.getActions()).toMatchSnapshot();
  });
});
