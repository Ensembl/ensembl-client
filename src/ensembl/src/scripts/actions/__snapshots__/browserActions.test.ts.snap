// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`browserActions dispaches the correct actions and payloads 1`] = `
Array [
  Object {
    "meta": undefined,
    "payload": undefined,
    "type": "browser/toggle-track-panel",
  },
  Object {
    "meta": undefined,
    "payload": undefined,
    "type": "browser/toggle-drawer",
  },
  Object {
    "meta": undefined,
    "payload": undefined,
    "type": "browser/open-drawer",
  },
  Object {
    "meta": undefined,
    "payload": undefined,
    "type": "browser/close-drawer",
  },
  Object {
    "meta": undefined,
    "payload": "track-1",
    "type": "browser/change-current-track",
  },
  Object {
    "meta": undefined,
    "payload": "main",
    "type": "browser/change-current-drawer-section",
  },
]
`;
